solidity/core/contracts/Inbox.sol

```diff
@@ -144,14 +140,19 @@ contract Inbox is IInbox, Version0, Common {
      * @param _messageHash keccak256 hash of the message
      */
     function _process(bytes calldata _message, bytes32 _messageHash) internal {
-        bytes29 _m = _message.ref(0);
         // ensure message was meant for this domain
-        require(_m.destination() == localDomain, "!destination");
+        require(_message.destination() == localDomain, "!destination");
```

Comment:
The nice thing with TypedMemView / `bytes memory` in general is that each time you pass bytes to a function, you're just passing a pointer to the bytes in memory which is super cheap. Tbh I'm still a little fuzzy on _exactly_ what is done with the `bytes calldata` when passing it to functions, but playing around with the contracts and observing the gas amounts, it seems like the entire `bytes calldata` is copied each time it's passed as a parameter. In other words I think each time we do something like `_message.destination()` or `_message.origin()` etc, the entire `bytes calldata _message` is copied, which gets expensive in this function when we're individually getting the different components of a message.

I don't think there's native `bytes memory` slicing (hence why something like TypedMemView exists ig lol), so I locally tried doing something like this so that the `bytes calldata _message` is only copied once when getting the message's fields via a `destructure()` call. Also note how I changed the _body from being `bytes memory` to `bytes calldata`:

In Message.sol:
```
function destructure(bytes calldata _message) internal pure returns (
    uint32 _origin,
    bytes32 _sender,
    uint32 _destination,
    address _recipient,
    bytes calldata _body
) {
    return (
        uint32(bytes4(_message[:4])),
        bytes32(_message[4:36]),
        uint32(bytes4( _message[36:40])),
        TypeCasts.bytes32ToAddress(bytes32(_message[40:72])),
        _message[72:]
    );
}
```

In `Inbox._process`:
```
function _process(bytes calldata _message, bytes32 _messageHash) internal {
    (
        uint32 _origin,
        bytes32 _sender,
        uint32 _destination,
        address _recipientAddress,
        bytes calldata _body
    ) = _message.destructure();

    // ensure message was meant for this domain
    require(_destination == localDomain, "!destination");

    // update message status as processed
    messages[_messageHash] = MessageStatus.Processed;
    IMessageRecipient _recipient = IMessageRecipient(
        _recipientAddress
    );
    _recipient.handle(
        _origin,
        _sender,
        _body
    );
    // emit process results
    emit Process(_messageHash);
}
```

I tested the gas used by logging the receipt of the `process` tx here: https://github.com/abacus-network/abacus-monorepo/blob/main/solidity/core/test/inbox.test.ts#L114

Without `destructure` the gas used was `85897`, with `destructure` the gas used was `85519`, so ~375 gas in savings. (If you're curious, `destructure` returning the body as `bytes memory` resulted in 100 gas savings, and changing that to `bytes calldata` gave the additional 275 in savings).

Even though this is modest, the `message` used in that test is pretty much as small as it gets bc the body is 2 bytes:
```
> sdk.parseMessage('0x000003e8000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb9226600000bb800000000000000000000000073511669fd4de447fed18bb79bafeac93ab7f31f1234')
{
  origin: 1000,
  sender: '0x000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266',
  destination: 3000,
  recipient: '0x00000000000000000000000073511669fd4de447fed18bb79bafeac93ab7f31f',
  body: '0x1234'
}
```
Because the body can get much larger I could see the gas used by all those `bytes calldata` copies getting much higher, so imo it's worth moving to something like `destructure` to keep costs as low as possible.

Separately, I think this does kinda highlight why TypedMemView is nice for a consuming Abacus app-- they can cheaply parse their message by doing individual calls like like `_message.someBodyField()` and `_message.anotherBodyField()` instead of needing to do something like `(bytes32 someBodyField, bytes32 anotherBodyField) = _message.destructureBody()` to save gas

Reply:
I added a `destructure` and `destructureAddresses`, however the latter isn't really useful if we always  want to allow the application to process non-evm origin or target domains